src/main/setupInvestHandlers.js

```diff
@@ -112,50 +102,66 @@ export function setupInvestRunHandlers(investExe) {
     ];
     logger.debug(`set to run ${cmdArgs}`);
     let investRun;
+    let investLogfile;
+    const env = {
+      PATH: path.dirname(investExe),
+      PYTHONUNBUFFERED: 'TRUE',
+    };
     if (process.platform !== 'win32') {
       investRun = spawn(path.basename(investExe), cmdArgs, {
-        env: { PATH: path.dirname(investExe) },
+        env: env,
         shell: true, // without shell, IOError when datastack.py loads json
         detached: true, // counter-intuitive, but w/ true: invest terminates when this shell terminates
       });
     } else { // windows
       investRun = spawn(path.basename(investExe), cmdArgs, {
-        env: { PATH: path.dirname(investExe) },
+        env: env,
         shell: true,
       });
     }
 
     // There's no general way to know that a spawned process started,
     // so this logic to listen once on stdout seems like the way.
-    investRun.stdout.once('data', async (data) => {
-      const logfile = await findMostRecentLogfile(args.workspace_dir);
-      // TODO: handle case when logfile is still undefined?
-      // Could be if some stdout is emitted before a logfile exists.
-      logger.debug(`invest logging to: ${logfile}`);
-      runningJobs[args.workspace_dir] = investRun.pid;
-      event.reply(`invest-logging-${channel}`, logfile);
-    });
+    // We need to store the PID to enable killing the task.
+    // And need to parse the logfile path from stdout.
+    const stdOutCallback = async (data) => {
+      if (!investLogfile) {
+        if (`${data}`.match('Writing log messages to')) {
+          investLogfile = `${data}`.split(' ').pop().trim();
+          runningJobs[jobID] = investRun.pid;
+          event.reply(`invest-logging-${jobID}`, investLogfile);
+        }
+      }
+      // we set python stdio to be unbuffered, so data here should
+      // only be one logger message at a time.
+      event.reply(
+        `invest-stdout-${jobID}`,
+        markupMessage(`${data}`, LOG_PATTERNS)
+      );
+    };
+    LOG_PATTERNS['invest-log-primary'] = new RegExp(pyModuleName);
+    investRun.stdout.on('data', stdOutCallback);
 
-    // Capture stderr to a string separate from the invest log
-    // so that it can be displayed separately when invest exits.
-    // And because it could actually be stderr emitted from the
-    // invest CLI or even the shell, rather than the invest model,
-    // in which case it's useful to logger.debug too.
-    investRun.stderr.on('data', (data) => {
+    const stdErrCallback = (data) => {
+      // The python Traceback for invest comes through stdout & stderr,
+      // So no need to merge those streams for the benefit of displaying
+      // a complete log.
```

Comment:
This was an unexpected realization. It's convenient, but do we know exactly why this happens. And also there's at least one case where the Traceback is *not* going through stdout. https://github.com/natcap/invest/issues/651

Reply:
Based on what I found while working on https://github.com/natcap/invest/pull/655, the intent of the way InVEST's logging is structured in a `--headless` run is that logging will be sent to 2 handlers: `StreamHandler(sys.stdout)` and a `FileHandler`.  So a `LOGGER.exception` call should appear in both, and an uncaught exception (the CLI doesn't handle an exception in `execute`) will be printed to `stderr` as you'd expect for a python program.  It does seem like there's some undefined behavior around whether an exception message would be written to `stdout` via logging before the application terminates and the exception is written to `stderr`.

The changes in https://github.com/natcap/invest/pull/655 _should_ make exception handling more consistent and should clarify to python that we want to log the exception in the CLI if `execute` raises one.